add dark mode

daaanny90 opened the pull request:
Implement dark mode.

The preferred theme remains the light one. Then, if the users's system scheme is dark, the dark mode will be activated too.

daaanny90 committed:
add dark mode

daaanny90 commented:
@davidenke we can discuss about the color variable naming... 😄 Still don't know which system is the best one. I went this time for numbered colours.

daaanny90 committed:
Add semver

davidenke reviewed (changes requested):
As discussed, we should declare grey scale colors as `hsl` and use the amount of light for naming it (`hsl(0deg 0% 25%)` -> `--ddjs-color-grey-25`).

daaanny90 committed:
Merge branch 'main' into dark-mode

daaanny90 committed:
WIP add theme switch

davidenke commented on src/components/icons/dark.icon.astro:
This is a good point to work through some SVG best practices :)
- I highly recommend to use https://jakearchibald.github.io/svgomg/ for prep
- drop class on SVG elements (especially if unused)
- most of the time `fill-rule` and `clip-rule` are unnecessary
- if `fill` is not a domain specific color, one should use `currentColor` to inherit the font color; this makes life easier to adopt icon colors to the styled environment or to just declare styles somewhere outside with the very declarative `color` prop

davidenke commented on src/components/nav.component.astro:
```diff
@@ -3,51 +3,52 @@ import InfoIcon from './icons/info.icon.astro';
+    
     background: var(--ddjs-color-4);
-
+    
```
Looks like your prettier is a bit off: those white spaces should be removed automagically...

daaanny90 replied:
Ops :) Maybe could be an idea to use git hooks to automate the prettify process?

davidenke replied:
Not a fan of that. It should be checked first in the pipeline, but as we're not novice developers anymore, we should trust our local setup...

daaanny90 replied:
never trust anyone 😈

Yep, it's ok for me

davidenke commented on src/components/theme-switch.component.astro:
```diff
@@ -0,0 +1,64 @@
+
+---
+<style lang="scss">
+  .light, .dark {
```
Look out: Those class names are very generic! As the styles would be scoped by Astros magic, your (global) js will not! Thus, using the class names `light` or `dark` can lead to misbehavior. I'd suggest to add a common `theme-switch` class, to be addressed with the styles for both in CSS (`.theme-switch`) and combined in JS (`.theme-switch.light`).

davidenke commented on src/components/theme-switch.component.astro:
```diff
@@ -0,0 +1,64 @@
+</theme-switch>
+
+<script>
+  class ThemeSwitch extends HTMLElement {
```
Using a web component would subsequently mean to bring the markup with it as well and to declare it as a library component separately. Then, Astro could do its island magic and we would have proper runtime code.
But that would mean to invest quite some more in tooling ([lit](https://lit.dev)!) and for now, we should just use plain ol' vanilla runtime JS as been done with the search...

davidenke commented on src/components/theme-switch.component.astro:
```diff
@@ -0,0 +1,64 @@
+      this.darkIcon!.classList.toggle('show');
+      this.lightIcon!.classList.toggle('show');
+      
+      const theme = (document.querySelector('html')!.getAttribute('theme') === 'dark') ? 'light' : 'dark'
```
I'd suggest to use the [DOM api to retrieve the document node](https://developer.mozilla.org/en-US/docs/Web/API/Document/documentElement).

davidenke commented on src/layouts/default.layout.scss:
```diff
@@ -5,18 +5,23 @@
+  @include base.theme;  
+}
+
+:root[theme="light"] {
```
So either using `document.querySelector('html')` along with `html[theme="light"]`, \
_or_ on root node `document.documentElement` with `:root[theme="light"]`.

We should avoid mixing those, as the API does not always resolve to `html` and is a bit implicit here (non-empty) - Please stick to one pattern.

daaanny90 replied:
I've changed the selector in `theme-switch.component.astro` to `document.documentElement`

davidenke commented on src/layouts/default.layout.scss:
```diff
@@ -5,18 +5,23 @@
   @include light.theme;
+}
 
-  @media(prefers-color-scheme: dark) {
```
I'd tend to say, that we should stick to the `prefers-color-scheme` as long as no theme has been explicitly set. Thus, the theme adopts to the users preferences (e.g. dynamically change scheme at runtime by the os). Once the theme is explicitly set, the attribute selector overrides this.

To keep it simple we would not offer an option to opt-out of an explicitly selected theme, unless you reload the page (or simply navigate).

So either way, this should be _opt-in_ and we should later on think about persisting this setting (e.g. in `sessionStorage` or `localStorage` or ones moms cookies...)

davidenke reviewed (changes requested):
We can review / work through this together if you want. This is a very common requirement and we should invest some time to create an elaborated solution as best practice pattern for the future ☀️.

daaanny90 committed:
Clean SVGs + Fix SVG color

daaanny90 committed:
Prettify file

daaanny90 committed:
Change theme switch to runtime

daaanny90 committed:
Change first load theme handling + persistent theme

Now in the first choice is made reading the preferred color scheme, and then the theme is explicit changed from the user with the theme-switch. The choiche is now persistent.

Commit:
Merge branch 'main' into dark-mode

davidenke commented on package.json:
The version will be changed in the ci process and should not be set manually.

davidenke commented on public/dark.svg:
As already mentioned, class names should be omitted, the `fill` should be set to `currentColor`.

davidenke commented on public/light.svg:
Omit class names.

davidenke commented on src/components/icons/dark.icon.astro:
Are the static SVGs now obsolete?

davidenke commented on src/components/icons/dark.icon.astro:
```diff
@@ -0,0 +1,10 @@
+---
+export interface Props {
```
I think the Props type must not be exported, if I remember the Astro docs correctly.

daaanny90 replied:
yes, you remeber correctly 😁

davidenke commented on src/components/icons/light.icon.astro:
See comments in dark icon.

davidenke commented on src/components/nav.component.astro:
```diff
@@ -22,7 +23,7 @@ const imprintLink = `${import.meta.env.BASE_URL}imprint`;
     justify-content: space-between;
 
-    background: var(--ddjs-color-white);
+    background: var(--ddjs-color-4);
```
Is there such a color? If so, the name seems to be incomplete. An index should describe some kind of color scope, e.g. `primary`, `grey` or `mono`.

davidenke commented on src/components/nav.component.astro:
```diff
@@ -3,6 +3,7 @@ import InfoIcon from './icons/info.icon.astro';
+import ThemeSwitch from './theme-switch.component.astro';
 
 const { pathname } = Astro.url;
 const imprintLink = `${import.meta.env.BASE_URL}imprint`;
```
Instead of using Nodes meta, we should stick to the props Astro is providing through it's API.

daaanny90 replied:
Done

davidenke commented on src/components/theme-switch.component.astro:
```diff
@@ -0,0 +1,66 @@
+
+<style lang="scss">
+  .theme-switch {
+    &.light,
```
If these are all states(?), this can be omitted in the selector.

davidenke commented on src/components/theme-switch.component.astro:
```diff
@@ -0,0 +1,66 @@
+  }
+</style>
+
+<div theme-switch>
```
As we do not have a custom element, this attribute won't be valid. Either use the `data-` prefix, or use an id / className.

davidenke commented on src/components/theme-switch.component.astro:
```diff
@@ -0,0 +1,66 @@
+  </span>
+</div>
+
+<script>
```
This script snippet we should discuss together... 😊

davidenke commented on src/layouts/default.layout.astro:
```diff
@@ -40,3 +40,11 @@ const { lang = 'en', title } = Astro.props;
 </html>
+
+<script>
+  const savedTheme = localStorage.getItem('theme');
```
This maybe is not the right place to initialize a preselected theme.

daaanny90 replied:
Why not?

davidenke commented on src/layouts/default.layout.scss:
```diff
@@ -7,38 +7,31 @@
+  @include base.theme;
+  @include light.theme;
+
+  @media (prefers-color-scheme: dark) {
```
For clarity, this should be conditional for the light theme as well.

davidenke commented on src/styles/themes/base.scss:
```diff
@@ -0,0 +1,32 @@
+  --ddjs-base-size: 8px;
+
+  /* Fonts */
+  --ddjs-font-default: 'Source Sans Pro';
```
This can be set from the sass var.

daaanny90 replied:
Can you explain a bit more pls? 🙂

davidenke reviewed (changes requested):
Let's have a look on this together... 🙃

daaanny90 committed:
Merge branch 'main' into dark-mode

daaanny90 committed:
Remove unused class and fix color

daaanny90 committed:
Remove export from props interface

daaanny90 committed:
Remove export from props interface

daaanny90 committed:
Update launch.json

daaanny90 committed:
Use Astro API instead of node

daaanny90 committed:
Use Astro API

daaanny90 committed:
Remove custom attribute and useless nested scss

daaanny90 committed:
Revert semver

daaanny90 committed:
Add explicit theme set
